src/tutorials/jbearing2.c

```diff
@@ -540,12 +542,18 @@ PetscErrorCode CallPermonAndCompareResults(Tao tao, void *ctx)
   ierr = QPGetSolutionVector(qp, &x_qp);CHKERRQ(ierr);
   ierr = TaoGetSolutionVector(tao,&x_tao);CHKERRQ(ierr);
   
-  /*Different of results from Snes and QP */
+  /* Different of results from Snes and QP */
+  ierr = VecNorm(user->B, NORM_2, &rhs_norm);CHKERRQ(ierr);
+  tao_diff_tol = 1e2 * rtol / rhs_norm;
```

Comment:
```suggestion
  tao_diff_tol = 1e1 * PetscMax(rtol*rhs_norm,atol); /* 10 times the tolerance for the default convergence test */
```
Throws an error when the difference is more than 10* the stopping tolerance and mentions to users what is the default tolerance.

Reply:
Oh, good point that rtol must be _multiplied_ with rhs_norm.

Reply:
OK, I will apply it and make sure all tests still pass.

Reply:
@jkruzik Unfortunately,
```
mpirun -n 2 ./jbearing2 -mx 100 -my 100
---- Journal Bearing Problem SHB-----
mx: 100,  my: 100,  ecc: 0.1 

Value of Tao are gatol = 1.000000e-08, grtol =  1.000000e-08, gttol = 0.000000e+00  
Norm of difference of results from TAO and QP = 4.807057e-06 > 1.000000e-07 = tolerance
[0]PETSC ERROR: --------------------- Error Message --------------------------------------------------------------
[0]PETSC ERROR: Petsc has generated inconsistent data
[0]PETSC ERROR: PERMON and TAO yield different results!
...
```
so I would like to return to multiplying by 1e2, although such big problem is not tested now.

Reply:
Actually, as you can see, even 1e-6 won't be big enough here to pass :-( Suggestions?

Reply:
I think we do not have the same stopping criteria:
```
Value of Tao are gatol = 1.000000e-08, grtol =  1.000000e-08, gttol = 0.000000e+00
```
According to [TaoSetTolerances()](https://www.mcs.anl.gov/petsc/petsc-current/docs/manualpages/Tao/TaoSetTolerances.html), grtol is relative to f(x) while, assuming x_0 = 0, gttol == rtol in QPS.

The same issue is in: https://github.com/permon/permon/blob/e2bcc13e1edf12299bb0d05d2fe67cd3074ee696/src/qps/impls/tao/qpstao.c#L188

Maybe you can try running with `-tao_converged_reason` to see what we are stopping on (I think it is grtol).

Reply:
Hm, maybe we could have the TAO stopping criterion (relative to f(x)) as an alternative stopping criterion in PERMON? It would be cool if could get exact 0.0 for TAO and QPSTAO with the same settings.

Reply:
Well, if we fix qpstao (say, ```TaoSetTolerances( tao, qps->atol, PETSC_MACHINE_EPSILON, qps->rtol)```) we should get diff very close to 0.

We could add something like ConvergenceTestTao() that would be the same as default ConvergenceTest in TAO.

Also, we could make an example with a custom convergence test with g(x)/f(x) stopping criterion. Wink, wink PR #40.

Reply:
> We could add something like ConvergenceTestTao() that would be the same as default ConvergenceTest in TAO.

Yes, it would be only valid for QPSTAO and would make use of all TAO tests because there are currently more of them than in PERMON.

This brings me to the main problem here: TAO is actually diverging for `-mx` and `-my` over ~30 due to `DIVERGED_MAXFCN`. It seems BLMVM has serious convergence problems with those bigger problems (not our problem, though).

So I suggest adding just checks that TAO and QPS both converge (otherwise it makes no sense to compare them) - will push shortly.

> Also, we could make an example with a custom convergence test with g(x)/f(x) stopping criterion. Wink, wink PR #40.

Yes, I actually have it already. I was even thinking about adding it to the library rather than just example. But the problem is we would have to make sure in all QPS implementation that the objective value is computed in every iteration, or would have an additional MatMult in every stopping criterion evaluation. I don't think we want either now.

So should I add it to this example for now? Within this PR?

Reply:
> So should I add it to this example for now? Within this PR?

If so, it would be better to get #40 merged first.

Reply:
> > We could add something like ConvergenceTestTao() that would be the same as default ConvergenceTest in TAO.
> 
> Yes, it would be only valid for QPSTAO and would make use of all TAO tests because there are currently more of them than in PERMON.
> 
I could only find the TaoConvergedDefault. More importantly, I would prefer QPSTAO to use by default our convergence test that is compatible with other solvers in PERMON (i.e., `TaoSetTolerances( tao, qps->atol, PETSC_MACHINE_EPSILON, qps->rtol)` where we need to check if TAO takes x_0 into account). In this way, we can freely swap solvers while having consistent stopping criteria.
> This brings me to the main problem here: TAO is actually diverging for `-mx` and `-my` over ~30 due to `DIVERGED_MAXFCN`. It seems BLMVM has serious convergence problems with those bigger problems (not our problem, though).
> 
Let us switch to TRON, both MPGP and TRON are stopping on ATOL=1e-8 and the difference between the solution is at worst 1e-8.
> So I suggest adding just checks that TAO and QPS both converge (otherwise it makes no sense to compare them) - will push shortly.
> 
Agreed.
> > Also, we could make an example with a custom convergence test with g(x)/f(x) stopping criterion. Wink, wink PR #40.
> 
> Yes, I actually have it already. I was even thinking about adding it to the library rather than just example. But the problem is we would have to make sure in all QPS implementation that the objective value is computed in every iteration, or would have an additional MatMult in every stopping criterion evaluation. I don't think we want either now.
> 
If an algorithm has computed a gradient (and it is saved into qps/ convergence ctx) we could often evaluate cost function cheaply using QPComputeObjectiveFromGradient().
> So should I add it to this example for now? Within this PR?

I think it would be better to do an example (a separate one, not in jbearing2) in #40, to show the use of QPSGetResidualNorm() and QPSGetIterationNumber() (which are almost universally needed due to removals in #40).

I am not sure how useful is a stopping criterion relative to f(x), because you can start with a large f(x) value (e.g. 1e+8) while the value near the solution can be close to zero but equally it could be very large e.g., -1e+8).

Reply:
> I could only find the TaoConvergedDefault. More importantly, I would prefer QPSTAO to use by default our convergence test that is compatible with other solvers in PERMON (i.e., TaoSetTolerances( tao, qps->atol, PETSC_MACHINE_EPSILON, qps->rtol) where we need to check if TAO takes x_0 into account). In this way, we can freely swap solvers while having consistent stopping criteria.

Definitely! I just meant that as an alternative, non-default.

Reply:
> If an algorithm has computed a gradient (and it is saved into qps/ convergence ctx) we could often evaluate cost function cheaply using QPComputeObjectiveFromGradient().

True.

Reply:
> I think it would be better to do an example (a separate one, not in jbearing2) in #40, to show the use of QPSGetResidualNorm() and QPSGetIterationNumber() (which are almost universally needed due to removals in #40).

OK, but please let's keep #40 concise and do that in a separate MR.

> I am not sure how useful is a stopping criterion relative to f(x), because you can start with a large f(x) value (e.g. 1e+8) while the value near the solution can be close to zero but equally it could be very large e.g., -1e+8).

At least it's standard in TAO. But it makes some sense. If your objective is very large, you are not interested in relatively small improvements of the residual, and vice versa. Maybe I would just add it as alternative convergence test which interprets rtol in a different (TAO) way rather than the default (KSP) way.

Reply:
> Let us switch to TRON, both MPGP and TRON are stopping on ATOL=1e-8 and the difference between the solution is at worst 1e-8.

OK, but I could still keep rtol=1e-6 here, no? The original 1e-8 sounds a bit like overkill to me.

Reply:
1e-6 is fine, but it should be set as `tao_gttol` while e.g., `tao_grtol=1e-50` (also there is no need to set `tao_gatol` to the default)

Reply:
Gotcha, KSP/PERMON rtol is indeed like TAO gttol. Kind of confusing, though.

Reply:
Moving the leftovers to Issue #43.